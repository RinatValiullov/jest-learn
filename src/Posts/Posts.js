import React, { Component } from "react";
import Input from "../Input/input";
import Pagination from "../Pagination/Pagination";
import Post from "../Post/Post";
import Select from "../Select/Select";
import Title from "../Title/Title";
import {
  BASE_PATH,
  HITS,
  PAGE_HITS,
  PAGE_PARAM,
  SEARCH_PARAM,
  SEARCH_PATH,
} from "./constants";
import "./posts.css";

class Posts extends Component {
  state = {
    searchQuery: "",
    result: {},
    hitsPerPage: 20,
    page: 0,
  };

  componentDidMount() {
    const { searchQuery, hitsPerPage, page } = this.state;
    this.fetchData(searchQuery, hitsPerPage, page);
  }

  fetchData = (searchQuery, hitsPerPage, page) => {
    fetch(
      `${BASE_PATH}${SEARCH_PATH}?${SEARCH_PARAM}${searchQuery}&${PAGE_HITS}${hitsPerPage}&${PAGE_PARAM}${page}`
    )
      .then((res) => res.json())
      .then((result) => this.setNews(result));
  };

  handleInputChange = ({ target: { value } }) => {
    this.setState({
      searchQuery: value,
    });
  };

  handleHitsChange = ({ target: { value } }) => {
    const { searchQuery } = this.state;

    this.setState(
      {
        hitsPerPage: +value,
        page: 0,
      },
      () => {
        this.fetchData(searchQuery, this.state.hitsPerPage, 0);
      }
    );
  };

  getSearch = ({ key }) => {
    if (key === "Enter") {
      const { searchQuery, hitsPerPage } = this.state;
      this.setState({
        page: 0,
      });
      this.fetchData(searchQuery, hitsPerPage, 0);
    }
  };

  setNews = (result) => {
    this.setState({ result });
  };

  handlePageChange = ({ target }) => {
    const btnType = target.getAttribute("data-name");
    let { page } = this.state;

    if (!isNaN(btnType)) {
      this.updatePage(+btnType);
    } else {
      if (btnType === "next") {
        this.updatePage(page + 1);
      }
      if (btnType === "prev") {
        this.updatePage(page - 1);
      }
    }
  };

  updatePage = (page) => {
    const { searchQuery, hitsPerPage } = this.state;
    this.setState(
      {
        page,
      },
      () => {
        this.fetchData(searchQuery, hitsPerPage, page);
      }
    );
  };

  render() {
    const { searchQuery, result, hitsPerPage } = this.state;
    const { hits = [], page, nbPages } = result;

    return (
      <div className="wrapper">
        <Title title="Hacker News" />
        <Select
          options={HITS}
          handleChange={this.handleHitsChange}
          value={hitsPerPage}
        />
        {hits.length > 0 && (
          <Pagination
            onClick={this.handlePageChange}
            page={page}
            lastPage={nbPages}
          />
        )}
        <Input
          onKeyPress={this.getSearch}
          onChange={this.handleInputChange}
          value={searchQuery}
        />
        <ul className="newsList">
          {hits.map(
            ({
              author,
              created_at,
              num_comments,
              objectID,
              title,
              points,
              url,
            }) => (
              <Post
                key={objectID}
                author={author}
                created_at={created_at}
                num_comments={num_comments}
                title={title}
                points={points}
                url={url}
              />
            )
          )}
        </ul>
      </div>
    );
  }
}

export default Posts;
