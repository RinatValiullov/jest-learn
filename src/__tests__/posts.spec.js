import React from "react";
import Posts from "../Posts/Posts";

describe("Posts component", () => {
  it("should render Post component", () => {
    const component = render(<Posts />);
    expect(component).toMatchSnapshot();
  });
});
