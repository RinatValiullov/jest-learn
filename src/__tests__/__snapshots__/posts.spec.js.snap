// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts component should render Post component 1`] = `
<div
  class="wrapper"
>
  <h1
    class="title"
  >
    Hacker News
  </h1>
  <div
    class="selectWrapper"
  >
    <select>
      <option
        value="10"
      >
        10
      </option>
      <option
        selected=""
        value="20"
      >
        20
      </option>
      <option
        value="40"
      >
        40
      </option>
      <option
        value="50"
      >
        50
      </option>
    </select>
    <span
      class="selectText"
    >
      per page
    </span>
  </div>
  <div
    class="inputWrapper"
  >
    <i
      class="fas fa-search"
    />
    <input
      class="input"
      placeholder="Click on search"
      value=""
    />
  </div>
  <ul
    class="newsList"
  >
    <li
      class="post"
    >
      <div
        class="description"
      >
        <a
          class="title"
          href="//test.url"
        >
          Jest & Enzyme
        </a>
        <span
          class="text"
        >
          100 points
        </span>
        <span
          class="comments"
        >
          10 comments
        </span>
        <span
          class="date"
        >
          5/4/2020
        </span>
        <span
          class="author"
        >
          Yauhen
        </span>
      </div>
    </li>
    <li
      class="post"
    >
      <div
        class="description"
      >
        <a
          class="title"
          href="//test2121.url"
        >
          TypeScript Basics
        </a>
        <span
          class="text"
        >
          10 points
        </span>
        <span
          class="comments"
        >
          8 comments
        </span>
        <span
          class="date"
        >
          5/6/2020
        </span>
        <span
          class="author"
        >
          Stepan
        </span>
      </div>
    </li>
  </ul>
</div>
`;
